Sources/NIOCore/AsyncSequences/NIOBackPressuredAsyncSequence.swift

```diff
@@ -0,0 +1,1063 @@
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct NIOBackPressuredAsyncSequence<
+    Element,
+    Strategy: NIOBackPressuredAsyncSequenceStrategy,
+    Delegate: NIOBackPressuredAsyncSequenceDelegate
+>: NIOSendable {
+    /// Simple struct for the return type of ``NIOBackPressuredAsyncSequence/makeSourceAndSequence(backPressureStrategy:delegate:)``.
+    public struct NewSequence {
+        /// The source of the ``NIOBackPressuredAsyncSequence`` used to yield and finish.
+        public let source: Source
+        /// The actual sequence which should be passed to the consumer.
+        public let sequence: NIOBackPressuredAsyncSequence
+
+        @usableFromInline
+        /* fileprivate */ internal init(
+            source: Source,
+            sequence: NIOBackPressuredAsyncSequence
+        ) {
+            self.source = source
+            self.sequence = sequence
+        }
+    }
+
+    /// This class is needed to hook the deinit to observe once all references to the ``NIOBackPressuredAsyncSequence`` are dropped.
+    ///
+    /// If we get move-only types we should be able to drop this class and use the `deinit` of the ``AsyncIterator`` struct itself.
+    ///
+    /// - Important: This is safe to be unchecked ``Sendable`` since the `storage` is ``Sendable`` itself.
+    @usableFromInline
+    /* fileprivate */ internal final class InternalClass: @unchecked NIOSendable {
+        @usableFromInline
+        internal let storage: Storage
+
+        @inlinable
+        init(storage: Storage) {
+            self.storage = storage
+        }
+
+        @inlinable
+        deinit {
+            storage.sequenceDeinitialized()
+        }
+    }
+
+    @usableFromInline
+    /* private */ internal let internalClass: InternalClass
+
+    /// Initializes a new ``NIOBackPressuredAsyncSequence`` and a ``NIOBackPressuredAsyncSequence/Source``.
+    ///
+    /// - Important: This method returns a tuple containing a ``NIOBackPressuredAsyncSequence/Source`` and
+    /// a ``NIOBackPressuredAsyncSequence``. The source MUST be held by the caller and
+    /// used to signal new elements or finish. The sequence MUST be passed to the actual subscriber and MUST NOT be held by the
+    /// caller. This is due to the fact that deiniting the sequence is used as part of a trigger to terminate the underlying source.
+    ///
+    /// - Parameters:
+    ///   - backPressureStrategy: The back-pressure strategy of the sequence.
+    ///   - delegate: The delegate of the sequence
+    /// - Returns: A ``NIOBackPressuredAsyncSequence/Source`` and a ``NIOBackPressuredAsyncSequence``.
+    @inlinable
+    public static func makeSourceAndSequence(
+        backPressureStrategy: Strategy,
+        delegate: Delegate
+    ) -> NewSequence {
+        let sequence = Self(
+            backPressureStrategy: backPressureStrategy,
+            delegate: delegate
+        )
+        let source = Source(storage: sequence.internalClass.storage)
```

Comment:
Can we add a file private getter to `sequence` to return the storage so we don't need two accessors here?